File: bindings/clr/XMLReader.cpp
//  (C) Copyright Johan Ditmar, Karel Hruda, Paul O'Neill & Luke Stedman 2009.
//  Use, modification and distribution are subject to the Boost Software License,
//  Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
//  http://www.boost.org/LICENSE_1_0.txt).

#if defined(_MSC_VER)
#    pragma warning(disable:4503 4996)
#endif

#include "XMLReader.hpp"
#include "VariantException.hpp"
#include "StringTranslator.hpp"

namespace protean { namespace clr {

    XMLReader::XMLReader(System::IO::TextReader^ reader, XMLMode mode) :
        m_mode(mode),
        m_reader(new boost::iostreams::stream<TextReader>(reader))
    {
		m_entityReaders = gcnew System::Collections::Generic::Dictionary<System::String^,System::IO::TextReader^>();
	}

    XMLReader::XMLReader(System::IO::TextReader^ reader) :
        m_mode(XMLMode::Default),
        m_reader(new boost::iostreams::stream<TextReader>(reader))
    {
		m_entityReaders = gcnew System::Collections::Generic::Dictionary<System::String^,System::IO::TextReader^>();
	}

    XMLReader::~XMLReader()
    {
        this->!XMLReader();
    }

    XMLReader::!XMLReader()
    {
        delete m_reader;
    }

    Variant^ XMLReader::Read()
    {
        BEGIN_TRANSLATE_ERROR();

        Variant^ result = gcnew Variant();

        protean::xml_reader reader(*m_reader, (int)m_mode);
		if (m_entityPath!=nullptr)
		{
			reader.set_entity_path(StringTranslator(m_entityPath).c_str());
		}
		if (m_externalSchema!=nullptr)
		{
			reader.set_external_schema(StringTranslator(m_externalSchema).c_str());
		}

		std::vector<boost::shared_ptr<boost::iostreams::stream<TextReader>>> streams; // used for clearing up
		for each(System::Collections::Generic::KeyValuePair<System::String^, System::IO::TextReader^> kvp in m_entityReaders)
		{
			std::string id(StringTranslator(kvp.Key).c_str());
			boost::shared_ptr<boost::iostreams::stream<TextReader>> stream(new boost::iostreams::stream<TextReader>(kvp.Value));
			streams.push_back(stream);

			reader.add_entity_stream(id, *stream);
		}

        reader >> result->get_internals();

        return result;
        
        END_TRANSLATE_ERROR();
    }

    void XMLReader::SetEntityPath(System::String^ path)
	{
		m_entityPath = path;
	}
    void XMLReader::SetExternalSchema(System::String^ id)
	{
		m_externalSchema = id;
	}
    void XMLReader::AddEntityReader(System::String^ id, System::IO::TextReader^ reader)
	{
		m_entityReaders->Add(id, reader);
	}


}} // protean::clr

File: bindings/cs/XMLReader.cs
//  (C) Copyright Johan Ditmar, Karel Hruda, Paul O'Neill & Luke Stedman 2009.
//  Use, modification and distribution are subject to the Boost Software License,
//  Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
//  http://www.boost.org/LICENSE_1_0.txt).

using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;

namespace protean {

    public class XMLReader
    {

        public XMLReader(System.IO.TextReader stream, XMLMode mode, IVariantObjectFactory factory, bool validateXsd, System.IO.TextReader xsdStream)
        {
            if ((mode & XMLMode.Preserve) != 0)
            {
                throw new VariantException("Preserve parser has not been implemented yet");
            }
            else
            {
                m_parser = new XMLDefaultParser(stream, mode, factory, xsdStream, validateXsd);
            }
        }

        public Variant Read()
        {
            m_parser.Parse();
            return m_parser.Read();
        }

        public static Variant Create(
            System.IO.TextReader stream, XMLMode mode = XMLMode.Default, IVariantObjectFactory factory = null,
            bool validateXsd = true, System.IO.TextReader xsdStream = null)
        {
            XMLReader reader = new XMLReader(stream, mode, factory, validateXsd, xsdStream);
            return reader.Read();
        }

        public static Variant FromString(string xml, XMLMode mode, IVariantObjectFactory factory)
        {
            using (System.IO.StringReader ms = new System.IO.StringReader(xml))
            {
                return Create(stream: ms, mode: mode, factory: factory);
            }
        }

        public static Variant FromString(string xml)
        {
            return FromString(xml, XMLMode.Default, null);
        }

        private XMLParserBase m_parser;
    }

} // protean
